### PR TITLE
Expose security context to any Hystrix command.

### PR DESCRIPTION
**Pull request by daniellavoie**
Reworked PR from #1054. This feature let Hystrix command access Spring SecurityContext even when using `THREAD` isolation strategy.

Fix issue gh-1054

### CODE REVIEW

**Comment by dsyer** [comment]
LGTM. If we are going to schedule Camden this summer I think we should put this on master (1.2.0).

**Comment by daniellavoie** [comment]
Glad I could help !

File: spring-cloud-netflix-core/pom.xml [hunk 1]
```diff
@@ -36,6 +36,16 @@
 		</dependency>
 		<dependency>
 			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-security</artifactId>
+			<optional>true</optional>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-test</artifactId>
+			<optional>true</optional>
```
Comment (spencergibb) [inline spring-cloud-netflix-core/pom.xml]: Why did you change this from `scope:test` to `optional:true`?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixRequestContextEnablerFilter.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.hystrix.strategy.concurrency.HystrixRequestContext;
+
+/**
+ * @author Daniel Lavoie
+ */
+public class HystrixRequestContextEnablerFilter implements Filter {
+	private static Logger LOGGER = LoggerFactory
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixRequestContextEnablerFilter.java]: Should be commons logging.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+@Configuration
+@Conditional(HystrixSecurityCondition.class)
+@ConditionalOnClass({ Hystrix.class, SecurityContext.class })
+public class HystrixSecurityAutoConfiguration {
+	@Autowired(required = false)
+	private CustomCommandExecutionHook customCommandExecutionHook;
+
+	@PostConstruct
+	public void init() {
+		HystrixPlugins.reset();
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java]: This worries me a bit. If the user has set other plugins, they will be unceremoniously reset.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixRequestContextEnablerFilter.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.hystrix.strategy.concurrency.HystrixRequestContext;
+
+/**
+ * @author Daniel Lavoie
+ */
+public class HystrixRequestContextEnablerFilter implements Filter {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixRequestContextEnablerFilter.java]: `HystrixRequestContextFilter`? I don't think `Enabler` helps.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/SecurityContextRegistratorCommandHook.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+			.getLogger(SecurityContextRegistratorCommandHook.class);
+
+	private CustomCommandExecutionHook customCommandHook;
+
+	public SecurityContextRegistratorCommandHook(CustomCommandExecutionHook customCommandHook) {
+		this.customCommandHook = customCommandHook;
+	}
+
+	@Override
+	public <T> void onRunStart(HystrixCommand<T> commandInstance) {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/SecurityContextRegistratorCommandHook.java]: `onRunStart` is deprecated, use `onExecutionStart`.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/ZuulProperties.java [hunk 6]
```diff
@@ -48,7 +48,7 @@
 	 * duplicated if the proxy and the backend are secured with Spring. By default they
 	 * are added to the ignored headers if Spring Security is present.
 	 */
-	private static final List<String> SECURITY_HEADERS = Arrays.asList("Pragma",
+	public static final List<String> SECURITY_HEADERS = Arrays.asList("Pragma",
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/ZuulProperties.java]: Why did you change this?

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/filters/ZuulPropertiesTests.java [hunk 7]
```diff
@@ -46,7 +46,8 @@ public void teardown() {
 
 	@Test
 	public void defaultIgnoredHeaders() {
-		assertTrue(this.zuul.getIgnoredHeaders().isEmpty());
+		assertTrue(this.zuul.getIgnoredHeaders()
+				.containsAll(ZuulProperties.SECURITY_HEADERS));
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/filters/ZuulPropertiesTests.java]: Again, why was this changed?

File: spring-cloud-netflix-core/src/test/resources/application.yml [hunk 8]
```diff
@@ -30,6 +30,9 @@ badClients:
   ribbon:
     MaxAutoRetriesNextServer: 10
     ReadTimeout: 200
+username:
+  ribbon:
+    listOfServers: localhost:${local.server.port}
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/resources/application.yml]: Does this work? If so, it surprises me because I didn't think it was available in time.

File: spring-cloud-netflix-core/pom.xml (hunk 1 above)
Comment (daniellavoie) [inline spring-cloud-netflix-core/pom.xml]: Because the dependencies are required by the AutoConfiguration on compilation.

**Comment by spencergibb** [comment]
How about a rebase from master as well.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixRequestContextEnablerFilter.java (hunk 2 above)
Comment (daniellavoie) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixRequestContextEnablerFilter.java]: Sure, I'll fix it.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/ZuulProperties.java (hunk 6 above)
Comment (daniellavoie) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/ZuulProperties.java]: Adding Spring Security changes the headers injected by Zuul. `ZuulPropertiesTests.defaultIngoredHeaders` started to fail for this reason. I updated the tests to expect the headers declared with `ZuulProperties.SECURITY_HEADERS`.

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/filters/ZuulPropertiesTests.java (hunk 7 above)
Comment (daniellavoie) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/filters/ZuulPropertiesTests.java]: See my remark on `ZuulProperties.java`.

**Comment by daniellavoie** [comment]
Thanks for the review ! I'll rebase and work on those remarks this weekend.

File: spring-cloud-netflix-core/src/test/resources/application.yml (hunk 8 above)
Comment (daniellavoie) [inline spring-cloud-netflix-core/src/test/resources/application.yml]: I got surprised too. But according to my unit tests it does. I'll double check when I'll rework the PR.

File: spring-cloud-netflix-core/pom.xml (hunk 1 above)
Comment (spencergibb) [inline spring-cloud-netflix-core/pom.xml]: That shouldn't be the case. If it is, something is broken.

Comment (spencergibb) [inline spring-cloud-netflix-core/pom.xml]: I pulled your branch and tried reverting this particular change and it worked fine, so please revert.

**Comment by spencergibb** [comment]
Thanks for all your work @daniellavoie!

File: spring-cloud-netflix-core/pom.xml (hunk 1 above)
Comment (daniellavoie) [inline spring-cloud-netflix-core/pom.xml]: Ah ! I miss read your review. I though you were asking why `spring-boot-starter-security` changed to optional:true. Changing `spring-boot-starter-test` to optional is indeed an error. I'll fix it.

**Comment by daniellavoie** [comment]
Hi @spencergibb ! What is the target date for 1.2.0.M1 ? I'm currently in vacation so I finally have some free time to finish my work on this PR.

**Comment by spencergibb** [comment]
After spring one, so August unless @dsyer says otherwise. I'm away from work as well.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java (hunk 3 above)
Comment (daniellavoie) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java]: Ok, I came around it. I'm calling reset because the test suite registers a dummy command execution hooks and blocks my tests since multiple hooks can't be registered. Since only one hook may be registered, I'm forcing the reset so my test can pass.

I am going to move this reset to my integration test. Users will simply have to understand that they won't be able to register a hook as they have the custom command hook mechanism. How do you feel with that @dsyer ?

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java]: Makes sense. You'll have to note that in documentation.

Comment (dsyer) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java]: We have a custom hook in sleuth. Probably need to look at that and make sure there is a way to add both.

Comment (daniellavoie) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java]: Thanks for the heads up @dsyer . 

Sleuth is actually calling `HystrixPlugins.reset` through `SleuthHystrixConcurrencyStrategy`, but @marcingrzejszczak is doing it properly. He keeps a reference of any existing custom plugin and re registers them after the reset. Sleuth is only interested in overwriting the concurrency strategy. I'll do the same in my own class before calling reset and throw an explanatory runtime exception in case the Command Hook is already defined. Both Hyxtris and Sleuth won't be impacted by any race condition.

**Comment by pivotal-issuemaster** [comment]
@daniellavoie Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=1093)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=1093) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/faq) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@daniellavoie Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=1093)!

**Comment by daniellavoie** [comment]
Got hit by travis mysteries once again. Build is fine now !

In the end, I decided to not implement some fancy runtime detection of command hook overwrite. I just stated clearly in the doc that any command hook would be overwritten and that the custom hook mechanism is available.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/SecurityContextRegistratorCommandHook.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+			if (LOGGER.isTraceEnabled())
+				LOGGER.trace("Injecting security context.");
+
+			SecurityContextHolder.setContext(context);
+		}
+
+		if (customCommandHook != null)
+			customCommandHook.onExecutionStart(commandInstance);
+	}
+}
```
Comment (pukkaone) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/SecurityContextRegistratorCommandHook.java]: I just started using Hystrix, so this question may be a result of my ignorance. I'm curious why you implemented a command execution hook instead of a concurrency strategy that overrides the [wrapCallable](https://github.com/Netflix/Hystrix/wiki/Plugins#concurrencystrategy) method. For example:
```java
import com.netflix.hystrix.strategy.concurrency.HystrixConcurrencyStrategy;
import java.util.concurrent.Callable;
import org.springframework.security.concurrent.DelegatingSecurityContextCallable;

public class SecurityContextConcurrencyStrategy extends HystrixConcurrencyStrategy {
    @Override
    public <T> Callable<T> wrapCallable(Callable<T> callable) {
        return super.wrapCallable(new DelegatingSecurityContextCallable(callable));
    }
}
```
After the executing the callable, `DelegatingSecurityContextCallable` clears the security context from the thread.

Comment (daniellavoie) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/SecurityContextRegistratorCommandHook.java]: Your approach sounds interresting. I lost access to internet. I'll check out if the DelegatingSecurityContextCallable is more elegant than the command hook as soon the router recovers. Seems like he is on a hangover.

**Comment by daniellavoie** [comment]
Ok, I've turned this PR around by using the `HystrixConcurrencyStrategy` plugin rather than the Command Hook. Result is less code and a better handling of the security context by using `DelegatingSecurityContextCallable`. Extension is still available in a friendlier manner simply by declaring a `HystrixConcurrencyStrategy` bean. Spring Cloud will detect it within the Spring Context and wrapping it inside it's own plugin. Doc has been adapted.

Feel free to share your thoughts on that rewrite. Thanks @pukkaone.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		HystrixMetricsPublisher metricsPublisher = HystrixPlugins.getInstance()
+				.getMetricsPublisher();
+		HystrixPropertiesStrategy propertiesStrategy = HystrixPlugins.getInstance()
+				.getPropertiesStrategy();
+		HystrixCommandExecutionHook commandExecutionHook = HystrixPlugins.getInstance()
+				.getCommandExecutionHook();
+
+		HystrixPlugins.reset();
+
+		// Registers existing plugins excepts the Command Hook.
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java]: Comment now out of date?

**Comment by spencergibb** [comment]
I like the update. @dsyer or @ryanjbaxter any bandwidth to get this in before M1?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java (hunk 10 above)
Comment (daniellavoie) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java]: Yeap ! Will fix it right away :)

**Comment by daniellavoie** [comment]
I'll rebase this PR in the evening.

**Comment by daniellavoie** [comment]
@ryanjbaxter Wouhooo, thanks for merging. Finally I can stop maintaining that branch :) Let me know if you need any help on any other 1.2.0.M1 issue.
